pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -191,25 +242,66 @@ func (l *ListOptionIndexer) afterDelete(key string, tx db.TXClient) error {
 	return nil
 }
 
+func (l *ListOptionIndexer) deleteLabels(key string, tx db.TXClient) error {
+	err := tx.StmtExec(tx.Stmt(l.deleteLabelsStmt), key)
+	if err != nil {
+		return &db.QueryError{QueryString: l.deleteLabelsQuery, Err: err}
+	}
+	return nil
+}
+
 // ListByOptions returns objects according to the specified list options and partitions.
 // Specifically:
 //   - an unstructured list of resources belonging to any of the specified partitions
 //   - the total number of resources (returned list might be a subset depending on pagination options in lo)
 //   - a continue token, if there are more pages after the returned one
 //   - an error instead of all of the above if anything went wrong
 func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, partitions []partition.Partition, namespace string) (*unstructured.UnstructuredList, int, string, error) {
+	queryInfo, err := l.constructQuery(lo, partitions, namespace, db.Sanitize(l.GetName()))
+	if err != nil {
+		return nil, 0, "", err
+	}
+	return l.executeQuery(ctx, queryInfo)
+}
+
+// QueryInfo is a helper-struct that is used to represent the core query and parameters when converting
+// a filter from the UI into a sql query
+type QueryInfo struct {
+	query       string
+	params      []any
+	countQuery  string
+	countParams []any
+	limit       int
+	offset      int
+}
+
+func (l *ListOptionIndexer) constructQuery(lo ListOptions, partitions []partition.Partition, namespace string, dbName string) (*QueryInfo, error) {
+	queryInfo := &QueryInfo{}
+
+	// First, what kind of filtering will we be doing?
 	// 1- Intro: SELECT and JOIN clauses
-	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, db.Sanitize(l.GetName()))
+	// There's a 1:1 correspondence between a base table and its _Fields table
+	// but it's possible that a key has no associated labels, so if we're doing a
+	// non-existence test on labels we need to do a LEFT OUTER JOIN
+	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, dbName)
```

Comment:
I see rows being duplicated when an object has multiple rows.

From [the data set shared before](https://drive.google.com/drive/folders/1xIRN5fZ4ye7OD6eHsO-tTOYkOD6RaT5J?usp=sharing):


Check that namespaces starting with `cattle` have multiple abels:
```sql
SELECT o.key, lt.label FROM "_v1_Namespace" o
    LEFT OUTER JOIN "_v1_Namespace_labels" lt ON o.key = lt.key
    WHERE o.key LIKE 'cattle%'
    ORDER BY o.key, lt.label;
```

Result: generally they do
| o.key                           | lt.key                               |
|---------------------------------|--------------------------------------|
| cattle-fleet-system             | field.cattle.io/projectId            |
| cattle-fleet-system             | kubernetes.io/metadata.name          |
| cattle-fleet-system             | objectset.rio.cattle.io/hash         |
| cattle-global-data              | cattle.io/creator                    |
| cattle-global-data              | field.cattle.io/projectId            |
| cattle-global-data              | kubernetes.io/metadata.name          |
| cattle-global-nt                | cattle.io/creator                    |
| cattle-global-nt                | field.cattle.io/projectId            |
| cattle-global-nt                | kubernetes.io/metadata.name          |
| cattle-impersonation-system     | authz.cluster.cattle.io/impersonator |
| cattle-impersonation-system     | cattle.io/creator                    |
| cattle-impersonation-system     | field.cattle.io/projectId            |
| cattle-impersonation-system     | kubernetes.io/metadata.name          |
| cattle-provisioning-capi-system | kubernetes.io/metadata.name          |
| cattle-system                   | field.cattle.io/projectId            |
| cattle-system                   | kubernetes.io/metadata.name          |
| cattle-ui-plugin-system         | kubernetes.io/metadata.name          |


Now, change the `SELECT` clause to actually return columns we are interested in this PR (and adding `o.key` because extremely long hexadecimal strings are hard for humans to read):

```sql
SELECT o.key, o.object, o.objectnonce, o.dekid FROM "_v1_Namespace" o
    LEFT OUTER JOIN "_v1_Namespace_labels" lt ON o.key = lt.key
    WHERE o.key LIKE 'cattle%'
    ORDER BY o.key, lt.label;
;
```

Results:
| o.key                           | o.object  | o.objectnonce | o.dekid |
|---------------------------------|-----------|---------------|---------|
| cattle-fleet-clusters-system    | 0x257F... |               | 0       |
| cattle-fleet-clusters-system    | 0x257F... |               | 0       |
| cattle-fleet-clusters-system    | 0x257F... |               | 0       |
| cattle-fleet-system             | 0x257F... |               | 0       |
| cattle-fleet-system             | 0x257F... |               | 0       |
| cattle-fleet-system             | 0x257F... |               | 0       |
| cattle-global-data              | 0x257F... |               | 0       |
| cattle-global-data              | 0x257F... |               | 0       |
| cattle-global-data              | 0x257F... |               | 0       |
| cattle-global-nt                | 0x257F... |               | 0       |
| cattle-global-nt                | 0x257F... |               | 0       |
| cattle-global-nt                | 0x257F... |               | 0       |
| cattle-impersonation-system     | 0x257F... |               | 0       |
| cattle-impersonation-system     | 0x257F... |               | 0       |
| cattle-impersonation-system     | 0x257F... |               | 0       |
| cattle-impersonation-system     | 0x257F... |               | 0       |
| cattle-provisioning-capi-system | 0x257F... |               | 0       |
| cattle-system                   | 0x257F... |               | 0       |
| cattle-system                   | 0x257F... |               | 0       |
| cattle-ui-plugin-system         | 0x257F... |               | 0       |


Can you reproduce those results? Could it be I am still missing something?

Reply:
Yes, I can definitely reproduce it, and here's the sql I used that gives clearer output (splitting the serialized object blob and splitting to show the start and end, because they all have the same start):
```
sqlite> SELECT o.key,
   ...>        substr(hex(o.object), 1, 10),
   ...>        substr(hex(o.object), length(hex(o.object)) - 10),
   ...>        substr(hex(o.objectnonce), 1, 10),
   ...> substr(hex(o.dekid), 1, 10) FROM "_v1_Namespace" o LEFT OUTER JOIN "_v1_Namespace_labels" lt ON o.key = lt.key WHERE lt.label like "%/%" ORDER BY o.key, lt.label;
cattle-fleet-clusters-system|257F030101|97374656D00||30
cattle-fleet-clusters-system|257F030101|97374656D00||30
cattle-fleet-clusters-system|257F030101|97374656D00||30
cattle-fleet-local-system|257F030101|E6574657300||30
cattle-fleet-local-system|257F030101|E6574657300||30
cattle-fleet-local-system|257F030101|E6574657300||30
…
```

So, yes, it needs a "DISTINCT" to remove the duplicates.

Reply:
OK, I am resolving this conversation now.

I'll be waiting on @richard-cox and colleagues to test various label filtering and keep an eye on whether any duplicates are ever returned (they should not).

Also as a note in the future: should this approach not work well (because DISTINCT adds some overhead) one possibility is to not JOIN labels at all and only look for them in subqueries in the WHERE clause (like it's done for the NOT IN case). This might or might not be beneficial for performance - but let's leave that optimization for later, if needed.

Thanks!